### PR TITLE
NODE-2409 New script serialization methods

### PR DESCRIPTION
**Pull request by ivan-mashonskiy**

### CODE REVIEW

File: lang/shared/src/main/scala/com/wavesplatform/lang/contract/serialization/ContractSerDeV2.scala [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  private def deserializeVerifierFunction(in: CodedInputStream): Either[String, VerifierFunction] = {
+    for {
+      a <- deserializeVerifiableAnnotation(in)
+      f <- deserializeDeclaration(in).map(_.asInstanceOf[FUNC])
+    } yield VerifierFunction(a, f)
+  }
+
+  private[lang] def deserializeList[A](in: CodedInputStream, df: CodedInputStream => Either[String, A]): Either[String, List[A]] = {
+    val len = in.readUInt32()
+    (1 to len).toList.traverse[Either[String, *], A](_ => df(in))
```
Comment (Karasiq) [inline lang/shared/src/main/scala/com/wavesplatform/lang/contract/serialization/ContractSerDeV2.scala]: `len` can be ill-formed and lead to allocation of Int.Max list

File: lang/shared/src/main/scala/com/wavesplatform/lang/contract/serialization/ContractSerDeV2.scala [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  private[lang] def deserializeOption[A](in: CodedInputStream, df: CodedInputStream => Either[String, A]): Either[String, Option[A]] = {
+    tryEi(in.readRawByte() > 0)
+      .flatMap {
+        case true  => df(in).map(_.some)
+        case false => Right(None)
```
Comment (Karasiq) [inline lang/shared/src/main/scala/com/wavesplatform/lang/contract/serialization/ContractSerDeV2.scala]:
```suggestion
    tryEi(in.readRawByte().toInt)
      .flatMap {
        case 1 => df(in).map(_.some)
        case 0 => Right(None)
        case _ => ???
```

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/compaction/ContractScriptCompactorV2.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      compact(dApp, State(0, dApp.meta.originalNames.toVector), saveMeta = false)
+    } else {
+      dApp
+    }
+  }
+
+  @tailrec
+  protected def createCompName(oldName: String, state: State[Vector[String]], dApp: DApp, saveMeta: Boolean): CompactionResult[String, Vector[String]] = {
+    if (saveMeta) {
+      val compNameIdx = state.originalNames.indexOf(oldName)
```
Comment (Karasiq) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/compaction/ContractScriptCompactorV2.scala]: indexOf is O(N) operation, so it's better to form a temporary hash map and operate with it.

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/compaction/ContractScriptCompactorV2.scala [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    } else {
+      val originalName = state.originalNames.applyOrElse[Int, String](nameToIdx(oldName), _ => oldName)
+      val name = if (originalName.isEmpty) oldName else originalName
+      CompactionResult(name, state)
+    }
+  }
+
+  protected def getReplacedName(oldName: String, state: State[Vector[String]], saveMeta: Boolean): String = {
+    if (saveMeta) {
+      val compNameIdx = state.originalNames.indexOf(oldName)
```
Comment (Karasiq) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/compaction/ContractScriptCompactorV2.scala]: indexOf is O(N) operation, so it's better to form a temporary hash map and operate with it.

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/ContractCompiler.scala [hunk 5]
```diff
@@ -218,8 +219,14 @@ object ContractCompiler {
 
         var resultDApp = DApp(metaWithErr._1.get, decs, callableFuncs, verifierFuncOptWithErr._1.get)
 
-        if (removeUnusedCode) resultDApp = ContractScriptCompactor.removeUnusedCode(resultDApp)
-        if (needCompaction) resultDApp = ContractScriptCompactor.compact(resultDApp)
+        if (removeUnusedCode) resultDApp = ContractScriptCompactorV1.removeUnusedCode(resultDApp)
+        if (needCompaction) {
+          if (version < V6) {
+            resultDApp = ContractScriptCompactorV1.compact(resultDApp)
```
Comment (Karasiq) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/ContractCompiler.scala]: No need to keep the old compaction algorithm.

**Review by Karasiq — CHANGES_REQUESTED** [review]

**Review by Karasiq — APPROVED** [review]

**Review by yvaskov — APPROVED** [review]

**Review by xrtm000 — APPROVED** [review]

**Review by phearnot — APPROVED** [review]

**Pull request merged**